subsys/bluetooth/mesh/delayable_msg.c

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <zephyr/sys/slist.h>
+
+#include <zephyr/net/buf.h>
+#include <zephyr/bluetooth/mesh.h>
+
+#include "msg.h"
+#include "access.h"
+
+#define LOG_LEVEL CONFIG_BT_MESH_ACCESS_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_mesh_delayable_msg);
+
+static void delayable_msg_handler(struct k_work *w);
+static bool push_msg_from_delayable_msgs(void);
+
+struct delayable_msg_chunk {
+	sys_snode_t node;
+	uint8_t data[CONFIG_BT_MESH_ACCESS_DELAYABLE_MSG_CHUNK_SIZE];
+};
+
+struct delayable_msg_ctx {
+	sys_snode_t node;
+	sys_slist_t chunks;
+	struct bt_mesh_msg_ctx ctx;
+	uint16_t src_addr;
+	const struct bt_mesh_send_cb *cb;
+	void *cb_data;
+	uint32_t fired_time;
+	size_t len;
+};
+
+static struct {
+	sys_slist_t busy_ctx;
+	struct k_work_delayable random_delay;
+} access_delayable_msg = {.random_delay = Z_WORK_DELAYABLE_INITIALIZER(delayable_msg_handler)};
+
+K_MEM_SLAB_DEFINE_STATIC(delayable_msg_chunks, sizeof(struct delayable_msg_chunk),
+			 CONFIG_BT_MESH_ACCESS_DELAYABLE_MSG_CHUNK_COUNT, 4);
+
+K_MEM_SLAB_DEFINE_STATIC(delayable_msgs_ctx, sizeof(struct delayable_msg_ctx),
```

Comment:
Adding any variable to the scalable array for self-managing is not efficient either. For example, customer makes an array size of 20, this is +16 bytes for nothing. In this aspect, the slab allocator is more efficient since it has permanent memory size consumption for self-management. I agree that Zephyr's slab allocator is not required since I do not use here any thread related routines. I will add slab-like approach to the delayable message mechanism. Moreover, it makes the complexity of the majority of linked-list access O(1) (except putting new context, it is O(n)). It has the advantage over the atomic bit approach where the procedure of getting free context will be O(n).

Reply:
Good question. Didn't try. I'll do it and add it here.

Reply:
mesh sample for nrf52840dk_nrf52840
```
with slab allocator
Memory region         Used Size  Region Size  %age Used
           FLASH:      213456 B         1 MB     20.36%
             RAM:       40502 B       256 KB     15.45%
        IDT_LIST:          0 GB         2 KB      0.00%

current version
Memory region         Used Size  Region Size  %age Used
           FLASH:      213496 B         1 MB     20.36%
             RAM:       40454 B       256 KB     15.43%
        IDT_LIST:          0 GB         2 KB      0.00%
```